src/editor/esphome-editor.ts

```diff
@@ -0,0 +1,289 @@
+wsLoc.protocol = "ws:";
+if (loc.protocol === "https:") {
+  wsLoc.protocol = "wss:";
+}
+const wsUrl = wsLoc.href;
+
+let editorActiveFilename: string | null = null;
+let editorActiveSecrets = false;
+let editorActiveWebSocket: WebSocket | null = null;
+let editorValidationScheduled = false;
+let editorValidationRunning = false;
+
+@customElement("esphome-editor")
+export class ESPHomeEditor extends LitElement {
+  private container: Ref<HTMLElement> = createRef();
+  editor?: monaco.editor.IStandaloneCodeEditor;
+  @property() theme?: string;
+  @property() language?: string;
+
+  @property() public configuration!: string;
+
+  static styles = css`
+    :host {
+      --editor-width: 100%;
+      --editor-height: 70vh;
+    }
+    main {
+      width: var(--editor-width);
+      height: var(--editor-height);
+    }
+  `;
+
+  public getValue() {
+    return this.editor!.getModel()?.getValue();
+  }
+
+  render() {
+    return html`
+      <style>
+        ${styles}
+        ${ESPHomeEditor.styles}
+      </style>
+      <div>start</div>
+      <div id="js-loading-indicator">
+        <div class="preloader-wrapper big active">
+          <div class="spinner-layer spinner-blue-only">
+            <div class="circle-clipper left">
+              <div class="circle"></div>
+            </div>
+            <div class="gap-patch">
+              <div class="circle"></div>
+            </div>
+            <div class="circle-clipper right">
+              <div class="circle"></div>
+            </div>
+          </div>
+        </div>
+      </div>
+      <div>end</div>
+      <main ${ref(this.container)}></main>
+    `;
+  }
+
+  firstUpdated() {
+    // @ts-ignore
+    self.MonacoEnvironment = {
+      getWorkerUrl: function (moduleId: string, label: string) {
+        // if (label === 'json') {
+        //   return './json.worker.bundle.js';
+        // }
+        // if (label === 'css' || label === 'scss' || label === 'less') {
+        //   return './css.worker.bundle.js';
+        // }
+        // if (label === 'html' || label === 'handlebars' || label === 'razor') {
+        //   return './html.worker.bundle.js';
+        // }
+        // if (label === 'typescript' || label === 'javascript') {
+        //   return './ts.worker.bundle.js';
+
+        return "./static/js/esphome/monaco-editor/esm/vs/editor/editor.worker.js";
+      },
+    };
+    this.editor = monaco.editor.create(this.container.value!, {
+      value: this.configuration,
+      language: "yaml",
+      theme: "dark",
+      automaticLayout: true,
+    });
+    // window
+    //   .matchMedia("(prefers-color-scheme: dark)")
+    //   .addEventListener("change", () => {
+    //     monaco.editor.setTheme(this.getTheme());
+    //   });
+    const filename = this.configuration;
+    editorActiveFilename = filename;
+    const isSecrets = filename === "secrets.yaml" || filename === "secrets.yml";
+    const filenameField = document.querySelector(
```

Comment:
If we need to query things, we should query local using `this.shadowRoot!.querySelector(...)`